node/src/Pos/GState/BlockExtra.hs

```diff
@@ -113,62 +113,68 @@ instance HasConfiguration => RocksBatchOp BlockExtraOp where
 ----------------------------------------------------------------------------
 
 foldlUpWhileM
-    :: forall a b ssc m r .
-    ( MonadBlockDB ssc m
+    :: forall a b m r .
+    ( MonadDBRead m
     , HasHeaderHash a
     )
-    => (Blund ssc -> m b)
-    -> a
-    -> ((Blund ssc, b) -> Int -> Bool)
-    -> (r -> b -> m r)
-    -> r
+    => (HeaderHash -> m (Maybe b)) -- ^ For each header we get b(lund)
+    -> a                           -- ^ We start iterating from it
+    -> (b -> Int -> Bool)          -- ^ Condition on b and depth
+    -> (r -> b -> m r)             -- ^ Conversion function
+    -> r                           -- ^ Starting value
     -> m r
-foldlUpWhileM morphM start condition accM init =
+foldlUpWhileM getData start condition accM init =
     loadUpWhileDo (headerHash start) 0 init
   where
     loadUpWhileDo :: HeaderHash -> Int -> r -> m r
-    loadUpWhileDo curH height !res = blkGetBlund curH >>= \case
+    loadUpWhileDo curH height !res = getData curH >>= \case
         Nothing -> pure res
-        Just x -> do
-            curB <- morphM x
+        Just someData -> do
             mbNextLink <- resolveForwardLink curH
-            if | not (condition (x, curB) height) -> pure res
+            if | not (condition someData height) -> pure res
                | Just nextLink <- mbNextLink -> do
-                     newRes <- accM res curB
+                     newRes <- accM res someData
                      loadUpWhileDo nextLink (succ height) newRes
-               | otherwise -> accM res curB
+               | otherwise -> accM res someData
 
--- Loads something from old to new.
+-- Loads something from old to new. foldlUpWhileM for (OldestFirst []).
 loadUpWhile
-    :: forall a b ssc m . (MonadBlockDB ssc m, HasHeaderHash a)
-    => (Blund ssc -> b)
+    :: forall a b m . (MonadDBRead m, HasHeaderHash a)
+    => (HeaderHash -> m (Maybe b))
     -> a
     -> (b -> Int -> Bool)
     -> m (OldestFirst [] b)
 loadUpWhile morph start condition = OldestFirst . reverse <$>
     foldlUpWhileM
-        (pure . morph)
+        morph
         start
-        (\b h -> condition (snd b) h)
+        condition
         (\l e -> pure (e : l))
         []
 
+-- | Return hashes loaded up. Basically a forward links traversal.
+loadHashesUpWhile
+    :: forall m a. (HasHeaderHash a, MonadDBRead m)
+    => a
+    -> (HeaderHash -> Int -> Bool)
+    -> m (OldestFirst [] HeaderHash)
+loadHashesUpWhile = loadUpWhile (pure . Just)
+
 -- | Returns headers loaded up.
 loadHeadersUpWhile
     :: (MonadBlockDB ssc m, HasHeaderHash a)
     => a
     -> (BlockHeader ssc -> Int -> Bool)
     -> m (OldestFirst [] (BlockHeader ssc))
-loadHeadersUpWhile start condition =
-    loadUpWhile (view blockHeader . fst) start condition
+loadHeadersUpWhile = loadUpWhile $ fmap (fmap $ view blockHeader) . blkGetBlock
```

Comment:
Is it possible to do this without `blkGetBlock`? Deserializing that is the heavy bit that we want to avoid.

Reply:
It is, i'll deliver a fix.